# Data Manipulation in R with dplyr

In [ ]:
https://rpubs.com/williamsurles/292547

## Load the dplyr and hflights package

In [2]:
# Load the dplyr package
library(dplyr)

# Load the hflights package
install.packages('hflights')
library(hflights)

# Call both head() and summary() on hflights
head(hflights)

package 'hflights' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\sxw17\AppData\Local\Temp\RtmpAjFCWW\downloaded_packages


Year Month DayofMonth DayOfWeek DepTime ArrTime UniqueCarrier FlightNum
5424 2011 1     1          6         1400    1500    AA            428      
5425 2011 1     2          7         1401    1501    AA            428      
5426 2011 1     3          1         1352    1502    AA            428      
5427 2011 1     4          2         1403    1513    AA            428      
5428 2011 1     5          3         1405    1507    AA            428      
5429 2011 1     6          4         1359    1503    AA            428      
     TailNum ActualElapsedTime ... ArrDelay DepDelay Origin Dest Distance
5424 N576AA  60                ... -10       0       IAH    DFW  224     
5425 N557AA  60                ...  -9       1       IAH    DFW  224     
5426 N541AA  70                ...  -8      -8       IAH    DFW  224     
5427 N403AA  70                ...   3       3       IAH    DFW  224     
5428 N492AA  62                ...  -3       5       IAH    DFW  224     
5429 N262AA  64                ...  -7      -1       IAH    DFW  224     
     TaxiIn TaxiOut Cancelled CancellationCode Diverted
5424 7      13      0                          0       
5425 6       9      0                          0       
5426 5      17      0                          0       
5427 9      22      0                          0       
5428 9       9      0                          0       
5429 6      13      0                          0

In [3]:
summary(hflights)

      Year          Month          DayofMonth      DayOfWeek        DepTime    
 Min.   :2011   Min.   : 1.000   Min.   : 1.00   Min.   :1.000   Min.   :   1  
 1st Qu.:2011   1st Qu.: 4.000   1st Qu.: 8.00   1st Qu.:2.000   1st Qu.:1021  
 Median :2011   Median : 7.000   Median :16.00   Median :4.000   Median :1416  
 Mean   :2011   Mean   : 6.514   Mean   :15.74   Mean   :3.948   Mean   :1396  
 3rd Qu.:2011   3rd Qu.: 9.000   3rd Qu.:23.00   3rd Qu.:6.000   3rd Qu.:1801  
 Max.   :2011   Max.   :12.000   Max.   :31.00   Max.   :7.000   Max.   :2400  
                                                                 NA's   :2905  
    ArrTime     UniqueCarrier        FlightNum      TailNum         
 Min.   :   1   Length:227496      Min.   :   1   Length:227496     
 1st Qu.:1215   Class :character   1st Qu.: 855   Class :character  
 Median :1617   Mode  :character   Median :1696   Mode  :character  
 Mean   :1578                      Mean   :1962                     
 3rd Qu.:1953  

## Explore the dataset

In [4]:
str(hflights)

'data.frame':	227496 obs. of  21 variables:
 $ Year             : int  2011 2011 2011 2011 2011 2011 2011 2011 2011 2011 ...
 $ Month            : int  1 1 1 1 1 1 1 1 1 1 ...
 $ DayofMonth       : int  1 2 3 4 5 6 7 8 9 10 ...
 $ DayOfWeek        : int  6 7 1 2 3 4 5 6 7 1 ...
 $ DepTime          : int  1400 1401 1352 1403 1405 1359 1359 1355 1443 1443 ...
 $ ArrTime          : int  1500 1501 1502 1513 1507 1503 1509 1454 1554 1553 ...
 $ UniqueCarrier    : chr  "AA" "AA" "AA" "AA" ...
 $ FlightNum        : int  428 428 428 428 428 428 428 428 428 428 ...
 $ TailNum          : chr  "N576AA" "N557AA" "N541AA" "N403AA" ...
 $ ActualElapsedTime: int  60 60 70 70 62 64 70 59 71 70 ...
 $ AirTime          : int  40 45 48 39 44 45 43 40 41 45 ...
 $ ArrDelay         : int  -10 -9 -8 3 -3 -7 -1 -16 44 43 ...
 $ DepDelay         : int  0 1 -8 3 5 -1 -1 -5 43 43 ...
 $ Origin           : chr  "IAH" "IAH" "IAH" "IAH" ...
 $ Dest             : chr  "DFW" "DFW" "DFW" "DFW" ...
 $ Distance        

## Section 2 - tbl, a special type of data.frame

tibble print adapts to the size of your window
glimpse gives you a more complete view of the tibble
if you don’t like it go back to data.frame and use str and head

## Convert data.frame to tibble

In [5]:
# Convert the hflights data.frame into a hflights tbl
hflights <- tbl_df(hflights)

# Display the hflights tbl
hflights

Year Month DayofMonth DayOfWeek DepTime ArrTime UniqueCarrier FlightNum
1      2011 1      1         6         1400    1500    AA            428      
2      2011 1      2         7         1401    1501    AA            428      
3      2011 1      3         1         1352    1502    AA            428      
4      2011 1      4         2         1403    1513    AA            428      
5      2011 1      5         3         1405    1507    AA            428      
6      2011 1      6         4         1359    1503    AA            428      
7      2011 1      7         5         1359    1509    AA            428      
8      2011 1      8         6         1355    1454    AA            428      
9      2011 1      9         7         1443    1554    AA            428      
10     2011 1     10         1         1443    1553    AA            428      
11     2011 1     11         2         1429    1539    AA            428      
12     2011 1     12         3         1419    1515    AA            428      
13     2011 1     13         4         1358    1501    AA            428      
14     2011 1     14         5         1357    1504    AA            428      
15     2011 1     15         6         1359    1459    AA            428      
16     2011 1     16         7         1359    1509    AA            428      
17     2011 1     17         1         1530    1634    AA            428      
18     2011 1     18         2         1408    1508    AA            428      
19     2011 1     19         3         1356    1503    AA            428      
20     2011 1     20         4         1507    1622    AA            428      
21     2011 1     21         5         1357    1459    AA            428      
22     2011 1     22         6         1355    1456    AA            428      
23     2011 1     23         7         1356    1501    AA            428      
24     2011 1     24         1         1356    1513    AA            428      
25     2011 1     25         2         1352    1452    AA            428      
26     2011 1     26         3         1353    1455    AA            428      
27     2011 1     27         4         1356    1458    AA            428      
28     2011 1     28         5         1359    1505    AA            428      
29     2011 1     29         6         1355    1455    AA            428      
30     2011 1     30         7         1359    1456    AA            428      
...    ...  ...   ...        ...       ...     ...     ...           ...      
227467 2011 12    6          2          752     857    WN            1628     
227468 2011 12    6          2          651     746    WN            2534     
227469 2011 12    6          2         1728    1932    WN             902     
227470 2011 12    6          2         1115    1315    WN            1167     
227471 2011 12    6          2         1753    1911    WN             513     
227472 2011 12    6          2         1222    1335    WN             581     
227473 2011 12    6          2          836     946    WN            1223     
227474 2011 12    6          2         1352    1749    WN            3085     
227475 2011 12    6          2         1850    2046    WN              39     
227476 2011 12    6          2          707     903    WN             424     
227477 2011 12    6          2         1335    1528    WN            1098     
227478 2011 12    6          2         1005    1158    WN            1403     
227479 2011 12    6          2         1723    1845    WN              33     
227480 2011 12    6          2         1139    1304    WN            1212     
227481 2011 12    6          2         2023    2109    WN             207     
227482 2011 12    6          2         1542    1637    WN             405     
227483 2011 12    6          2         1728    1825    WN             628     
227484 2011 12    6          2         1252    1349    WN             994     
227485 2011 12    6          2          857     955    WN

In [6]:
glimpse(hflights)

Observations: 227,496
Variables: 21
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15...
$ DayOfWeek         <int> 6, 7, 1, 2, 3, 4, 5, 6, 7, 1, 2, 3, 4, 5, 6, 7, 1...
$ DepTime           <int> 1400, 1401, 1352, 1403, 1405, 1359, 1359, 1355, 1...
$ ArrTime           <int> 1500, 1501, 1502, 1513, 1507, 1503, 1509, 1454, 1...
$ UniqueCarrier     <chr> "AA", "AA", "AA", "AA", "AA", "AA", "AA", "AA", "...
$ FlightNum         <int> 428, 428, 428, 428, 428, 428, 428, 428, 428, 428,...
$ TailNum           <chr> "N576AA", "N557AA", "N541AA", "N403AA", "N492AA",...
$ ActualElapsedTime <int> 60, 60, 70, 70, 62, 64, 70, 59, 71, 70, 70, 56, 6...
$ AirTime           <int> 40, 45, 48, 39, 44, 45, 43, 40, 41, 45, 42, 41, 4...
$ ArrDelay          <int> -10, -9, -8, 3, -3, -7, -1, -16, 44, 43, 29, 5, -...
$ DepDelay      

In [7]:
# Create the object carriers, containing only the UniqueCarrier variable of hflights
carriers <- hflights$UniqueCarrier
str(carriers)

 chr [1:227496] "AA" "AA" "AA" "AA" "AA" "AA" "AA" "AA" "AA" "AA" "AA" "AA" ...


## Changing labels of hflight, part 1 of 2

In [8]:
# Both the dplyr and hflights packages are loaded into workspace
lut <- c("AA" = "American", "AS" = "Alaska", "B6" = "JetBlue", "CO" = "Continental", 
         "DL" = "Delta", "OO" = "SkyWest", "UA" = "United", "US" = "US_Airways", 
         "WN" = "Southwest", "EV" = "Atlantic_Southeast", "F9" = "Frontier", 
         "FL" = "AirTran", "MQ" = "American_Eagle", "XE" = "ExpressJet", "YV" = "Mesa")

# Use lut to translate the UniqueCarrier column of hflights
hflights$UniqueCarrier <- lut[hflights$UniqueCarrier]

# Inspect the resulting raw values of your variables
glimpse(hflights)

Observations: 227,496
Variables: 21
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15...
$ DayOfWeek         <int> 6, 7, 1, 2, 3, 4, 5, 6, 7, 1, 2, 3, 4, 5, 6, 7, 1...
$ DepTime           <int> 1400, 1401, 1352, 1403, 1405, 1359, 1359, 1355, 1...
$ ArrTime           <int> 1500, 1501, 1502, 1513, 1507, 1503, 1509, 1454, 1...
$ UniqueCarrier     <chr> "American", "American", "American", "American", "...
$ FlightNum         <int> 428, 428, 428, 428, 428, 428, 428, 428, 428, 428,...
$ TailNum           <chr> "N576AA", "N557AA", "N541AA", "N403AA", "N492AA",...
$ ActualElapsedTime <int> 60, 60, 70, 70, 62, 64, 70, 59, 71, 70, 70, 56, 6...
$ AirTime           <int> 40, 45, 48, 39, 44, 45, 43, 40, 41, 45, 42, 41, 4...
$ ArrDelay          <int> -10, -9, -8, 3, -3, -7, -1, -16, 44, 43, 29, 5, -...
$ DepDelay      

## Changing labels of hflights, part 2 or 2

In [9]:
## The "E" in my dataset is "" for some reason so I have changed it here
table(hflights$CancellationCode)


            A      B      C      D 
224523   1202   1652    118      1 

In [10]:
hflights <- hflights %>%
  mutate(
    CancellationCode = ifelse(CancellationCode == "", "E", CancellationCode)
  )

table(hflights$CancellationCode)


     A      B      C      D      E 
  1202   1652    118      1 224523 

In [11]:
# Build the lookup table: lut
lut <- c("A" = "carrier",
         "B" = "weather",
         "C" = "FFA",
         "D" = "security",
         "E" = "not cancelled")

# Use the lookup table to create a vector of code labels. Assign the vector to the CancellationCode column of hflights
hflights$Code <- lut[hflights$CancellationCode]

# Inspect the resulting raw values of your variables
glimpse(hflights)

Observations: 227,496
Variables: 22
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15...
$ DayOfWeek         <int> 6, 7, 1, 2, 3, 4, 5, 6, 7, 1, 2, 3, 4, 5, 6, 7, 1...
$ DepTime           <int> 1400, 1401, 1352, 1403, 1405, 1359, 1359, 1355, 1...
$ ArrTime           <int> 1500, 1501, 1502, 1513, 1507, 1503, 1509, 1454, 1...
$ UniqueCarrier     <chr> "American", "American", "American", "American", "...
$ FlightNum         <int> 428, 428, 428, 428, 428, 428, 428, 428, 428, 428,...
$ TailNum           <chr> "N576AA", "N557AA", "N541AA", "N403AA", "N492AA",...
$ ActualElapsedTime <int> 60, 60, 70, 70, 62, 64, 70, 59, 71, 70, 70, 56, 6...
$ AirTime           <int> 40, 45, 48, 39, 44, 45, 43, 40, 41, 45, 42, 41, 4...
$ ArrDelay          <int> -10, -9, -8, 3, -3, -7, -1, -16, 44, 43, 29, 5, -...
$ DepDelay      

## Select and mutate

### Section 3 - The five verbs and select in more detail

5 verbs
select() which returns a subset of the columns,
filter() that is able to return a subset of the rows,
arrange() that reorders the rows according to single or multiple variables,
mutate() used to add columns from existing data,
summarise() which reduces each group to a single row by calculating aggregate measures.
verb focus
select and mutate manipulate variables
filter and arrange manipulate observations
summarize manipulates groups of observations

### Choosing is not losing! The select verb

In [12]:
# Print out a tbl with the four columns of hflights related to delay
select(hflights, ActualElapsedTime, AirTime, ArrDelay, DepDelay)

ActualElapsedTime AirTime ArrDelay DepDelay
1      60                40      -10       0      
2      60                45       -9       1      
3      70                48       -8      -8      
4      70                39        3       3      
5      62                44       -3       5      
6      64                45       -7      -1      
7      70                43       -1      -1      
8      59                40      -16      -5      
9      71                41       44      43      
10     70                45       43      43      
11     70                42       29      29      
12     56                41        5      19      
13     63                44       -9      -2      
14     67                47       -6      -3      
15     60                44      -11      -1      
16     70                41       -1      -1      
17     64                48       84      90      
18     60                42       -2       8      
19     67                46       -7      -4      
20     75                42       72      67      
21     62                47      -11      -3      
22     61                44      -14      -5      
23     65                40       -9      -4      
24     77                43        3      -4      
25     60                40      -18      -8      
26     62                40      -15      -7      
27     62                40      -12      -4      
28     66                46       -5      -1      
29     60                46      -15      -5      
30     57                39      -14      -1      
...    ...               ...     ...      ...     
227467  65                50       2      -3      
227468  55                43      -9      -4      
227469 244               231       7      23      
227470 240               225     -15       5      
227471  78                59      31      33      
227472  73                61     -10      -3      
227473  70                58      -9       1      
227474 177               163      59      72      
227475 176               157      71      70      
227476 176               162      -2      -3      
227477 173               159      18      20      
227478 173               163      13      15      
227479 202               192      70      78      
227480 205               193      -6      -1      
227481  46                38      29      43      
227482  55                43      12      12      
227483  57                39       5       3      
227484  57                43       9       7      
227485  58                44       0      -3      
227486 100                81      14      39      
227487 109                87     -10      -4      
227488 102                86     -12       1      
227489  95                83      -9      16      
227490 113               100      -4      -2      
227491 113                98       0       7      
227492 113                97      -9       8      
227493 107                94       4       7      
227494  79                61      -4      -3      
227495  76                64     -13      -4      
227496  73                59     -12       0

In [13]:
# Print out hflights, nothing has changed!
hflights

Year Month DayofMonth DayOfWeek DepTime ArrTime UniqueCarrier FlightNum
1      2011 1      1         6         1400    1500    American      428      
2      2011 1      2         7         1401    1501    American      428      
3      2011 1      3         1         1352    1502    American      428      
4      2011 1      4         2         1403    1513    American      428      
5      2011 1      5         3         1405    1507    American      428      
6      2011 1      6         4         1359    1503    American      428      
7      2011 1      7         5         1359    1509    American      428      
8      2011 1      8         6         1355    1454    American      428      
9      2011 1      9         7         1443    1554    American      428      
10     2011 1     10         1         1443    1553    American      428      
11     2011 1     11         2         1429    1539    American      428      
12     2011 1     12         3         1419    1515    American      428      
13     2011 1     13         4         1358    1501    American      428      
14     2011 1     14         5         1357    1504    American      428      
15     2011 1     15         6         1359    1459    American      428      
16     2011 1     16         7         1359    1509    American      428      
17     2011 1     17         1         1530    1634    American      428      
18     2011 1     18         2         1408    1508    American      428      
19     2011 1     19         3         1356    1503    American      428      
20     2011 1     20         4         1507    1622    American      428      
21     2011 1     21         5         1357    1459    American      428      
22     2011 1     22         6         1355    1456    American      428      
23     2011 1     23         7         1356    1501    American      428      
24     2011 1     24         1         1356    1513    American      428      
25     2011 1     25         2         1352    1452    American      428      
26     2011 1     26         3         1353    1455    American      428      
27     2011 1     27         4         1356    1458    American      428      
28     2011 1     28         5         1359    1505    American      428      
29     2011 1     29         6         1355    1455    American      428      
30     2011 1     30         7         1359    1456    American      428      
...    ...  ...   ...        ...       ...     ...     ...           ...      
227467 2011 12    6          2          752     857    Southwest     1628     
227468 2011 12    6          2          651     746    Southwest     2534     
227469 2011 12    6          2         1728    1932    Southwest      902     
227470 2011 12    6          2         1115    1315    Southwest     1167     
227471 2011 12    6          2         1753    1911    Southwest      513     
227472 2011 12    6          2         1222    1335    Southwest      581     
227473 2011 12    6          2          836     946    Southwest     1223     
227474 2011 12    6          2         1352    1749    Southwest     3085     
227475 2011 12    6          2         1850    2046    Southwest       39     
227476 2011 12    6          2          707     903    Southwest      424     
227477 2011 12    6          2         1335    1528    Southwest     1098     
227478 2011 12    6          2         1005    1158    Southwest     1403     
227479 2011 12    6          2         1723    1845    Southwest       33     
227480 2011 12    6          2         1139    1304    Southwest     1212     
227481 2011 12    6          2         2023    2109    Southwest      207     
227482 2011 12    6          2         1542    1637    Southwest      405     
227483 2011 12    6          2         1728    1825    Southwest      628     
227484 2011 12    6          2         1252    1349    Southwest      994     
227485 2011 12    6          2          857     955    So

In [14]:
# Print out the columns Origin up to Cancelled of hflights
select(hflights, Origin:Cancelled)

Origin Dest Distance TaxiIn TaxiOut Cancelled
1      IAH    DFW  224       7     13      0        
2      IAH    DFW  224       6      9      0        
3      IAH    DFW  224       5     17      0        
4      IAH    DFW  224       9     22      0        
5      IAH    DFW  224       9      9      0        
6      IAH    DFW  224       6     13      0        
7      IAH    DFW  224      12     15      0        
8      IAH    DFW  224       7     12      0        
9      IAH    DFW  224       8     22      0        
10     IAH    DFW  224       6     19      0        
11     IAH    DFW  224       8     20      0        
12     IAH    DFW  224       4     11      0        
13     IAH    DFW  224       6     13      0        
14     IAH    DFW  224       5     15      0        
15     IAH    DFW  224       6     10      0        
16     IAH    DFW  224      12     17      0        
17     IAH    DFW  224       8      8      0        
18     IAH    DFW  224       7     11      0        
19     IAH    DFW  224      10     11      0        
20     IAH    DFW  224       9     24      0        
21     IAH    DFW  224       6      9      0        
22     IAH    DFW  224       9      8      0        
23     IAH    DFW  224       7     18      0        
24     IAH    DFW  224       6     28      0        
25     IAH    DFW  224       7     13      0        
26     IAH    DFW  224       8     14      0        
27     IAH    DFW  224      12     10      0        
28     IAH    DFW  224       8     12      0        
29     IAH    DFW  224       7      7      0        
30     IAH    DFW  224       7     11      0        
...    ...    ...  ...      ...    ...     ...      
227467 HOU    MSY   302      4     11      0        
227468 HOU    MSY   302      5      7      0        
227469 HOU    OAK  1642      5      8      0        
227470 HOU    OAK  1642      6      9      0        
227471 HOU    OKC   419      4     15      0        
227472 HOU    OKC   419      3      9      0        
227473 HOU    OKC   419      4      8      0        
227474 HOU    PHL  1336      5      9      0        
227475 HOU    PHX  1020      4     15      0        
227476 HOU    PHX  1020      4     10      0        
227477 HOU    PHX  1020      4     10      0        
227478 HOU    PHX  1020      3      7      0        
227479 HOU    SAN  1313      3      7      0        
227480 HOU    SAN  1313      2     10      0        
227481 HOU    SAT   192      4      4      0        
227482 HOU    SAT   192      4      8      0        
227483 HOU    SAT   192      4     14      0        
227484 HOU    SAT   192      5      9      0        
227485 HOU    SAT   192      5      9      0        
227486 HOU    STL   687      4     15      0        
227487 HOU    STL   687     13      9      0        
227488 HOU    STL   687      4     12      0        
227489 HOU    STL   687      3      9      0        
227490 HOU    TPA   781      3     10      0        
227491 HOU    TPA   781      5     10      0        
227492 HOU    TPA   781      5     11      0        
227493 HOU    TPA   781      4      9      0        
227494 HOU    TUL   453      4     14      0        
227495 HOU    TUL   453      3      9      0        
227496 HOU    TUL   453      3     11      0

In [15]:
# Answer to last question: be concise!
select(hflights, 1:4, 12:21)

Year Month DayofMonth DayOfWeek ArrDelay DepDelay Origin Dest Distance
1      2011 1      1         6         -10       0       IAH    DFW  224     
2      2011 1      2         7          -9       1       IAH    DFW  224     
3      2011 1      3         1          -8      -8       IAH    DFW  224     
4      2011 1      4         2           3       3       IAH    DFW  224     
5      2011 1      5         3          -3       5       IAH    DFW  224     
6      2011 1      6         4          -7      -1       IAH    DFW  224     
7      2011 1      7         5          -1      -1       IAH    DFW  224     
8      2011 1      8         6         -16      -5       IAH    DFW  224     
9      2011 1      9         7          44      43       IAH    DFW  224     
10     2011 1     10         1          43      43       IAH    DFW  224     
11     2011 1     11         2          29      29       IAH    DFW  224     
12     2011 1     12         3           5      19       IAH    DFW  224     
13     2011 1     13         4          -9      -2       IAH    DFW  224     
14     2011 1     14         5          -6      -3       IAH    DFW  224     
15     2011 1     15         6         -11      -1       IAH    DFW  224     
16     2011 1     16         7          -1      -1       IAH    DFW  224     
17     2011 1     17         1          84      90       IAH    DFW  224     
18     2011 1     18         2          -2       8       IAH    DFW  224     
19     2011 1     19         3          -7      -4       IAH    DFW  224     
20     2011 1     20         4          72      67       IAH    DFW  224     
21     2011 1     21         5         -11      -3       IAH    DFW  224     
22     2011 1     22         6         -14      -5       IAH    DFW  224     
23     2011 1     23         7          -9      -4       IAH    DFW  224     
24     2011 1     24         1           3      -4       IAH    DFW  224     
25     2011 1     25         2         -18      -8       IAH    DFW  224     
26     2011 1     26         3         -15      -7       IAH    DFW  224     
27     2011 1     27         4         -12      -4       IAH    DFW  224     
28     2011 1     28         5          -5      -1       IAH    DFW  224     
29     2011 1     29         6         -15      -5       IAH    DFW  224     
30     2011 1     30         7         -14      -1       IAH    DFW  224     
...    ...  ...   ...        ...       ...      ...      ...    ...  ...     
227467 2011 12    6          2           2      -3       HOU    MSY   302    
227468 2011 12    6          2          -9      -4       HOU    MSY   302    
227469 2011 12    6          2           7      23       HOU    OAK  1642    
227470 2011 12    6          2         -15       5       HOU    OAK  1642    
227471 2011 12    6          2          31      33       HOU    OKC   419    
227472 2011 12    6          2         -10      -3       HOU    OKC   419    
227473 2011 12    6          2          -9       1       HOU    OKC   419    
227474 2011 12    6          2          59      72       HOU    PHL  1336    
227475 2011 12    6          2          71      70       HOU    PHX  1020    
227476 2011 12    6          2          -2      -3       HOU    PHX  1020    
227477 2011 12    6          2          18      20       HOU    PHX  1020    
227478 2011 12    6          2          13      15       HOU    PHX  1020    
227479 2011 12    6          2          70      78       HOU    SAN  1313    
227480 2011 12    6          2          -6      -1       HOU    SAN  1313    
227481 2011 12    6          2          29      43       HOU    SAT   192    
227482 2011 12    6          2          12      12       HOU    SAT   192    
227483 2011 12    6          2           5       3       HOU    SAT   192    
227484 2011 12    6          2           9       7       HOU    SAT   192    
227485 2011 12    6          2           0      -3       HOU    SAT   192    
227486 2011 12    6          

### Helper functions for variable selection

In [16]:
# Print out a tbl containing just ArrDelay and DepDelay
select(hflights, ends_with(c('Delay')))

ArrDelay DepDelay
1      -10       0      
2       -9       1      
3       -8      -8      
4        3       3      
5       -3       5      
6       -7      -1      
7       -1      -1      
8      -16      -5      
9       44      43      
10      43      43      
11      29      29      
12       5      19      
13      -9      -2      
14      -6      -3      
15     -11      -1      
16      -1      -1      
17      84      90      
18      -2       8      
19      -7      -4      
20      72      67      
21     -11      -3      
22     -14      -5      
23      -9      -4      
24       3      -4      
25     -18      -8      
26     -15      -7      
27     -12      -4      
28      -5      -1      
29     -15      -5      
30     -14      -1      
...    ...      ...     
227467   2      -3      
227468  -9      -4      
227469   7      23      
227470 -15       5      
227471  31      33      
227472 -10      -3      
227473  -9       1      
227474  59      72      
227475  71      70      
227476  -2      -3      
227477  18      20      
227478  13      15      
227479  70      78      
227480  -6      -1      
227481  29      43      
227482  12      12      
227483   5       3      
227484   9       7      
227485   0      -3      
227486  14      39      
227487 -10      -4      
227488 -12       1      
227489  -9      16      
227490  -4      -2      
227491   0       7      
227492  -9       8      
227493   4       7      
227494  -4      -3      
227495 -13      -4      
227496 -12       0

In [17]:
# Print out a tbl as described in the second instruction, using both helper functions and variable names
select(hflights, UniqueCarrier, ends_with(c('Num')), starts_with(c('Cancel')))

UniqueCarrier FlightNum TailNum Cancelled CancellationCode
1      American      428       N576AA  0         E               
2      American      428       N557AA  0         E               
3      American      428       N541AA  0         E               
4      American      428       N403AA  0         E               
5      American      428       N492AA  0         E               
6      American      428       N262AA  0         E               
7      American      428       N493AA  0         E               
8      American      428       N477AA  0         E               
9      American      428       N476AA  0         E               
10     American      428       N504AA  0         E               
11     American      428       N565AA  0         E               
12     American      428       N577AA  0         E               
13     American      428       N476AA  0         E               
14     American      428       N552AA  0         E               
15     American      428       N462AA  0         E               
16     American      428       N555AA  0         E               
17     American      428       N518AA  0         E               
18     American      428       N507AA  0         E               
19     American      428       N523AA  0         E               
20     American      428       N425AA  0         E               
21     American      428       N251AA  0         E               
22     American      428       N551AA  0         E               
23     American      428       N479AA  0         E               
24     American      428       N531AA  0         E               
25     American      428       N561AA  0         E               
26     American      428       N541AA  0         E               
27     American      428       N512AA  0         E               
28     American      428       N4UBAA  0         E               
29     American      428       N491AA  0         E               
30     American      428       N561AA  0         E               
...    ...           ...       ...     ...       ...             
227467 Southwest     1628      N435WN  0         E               
227468 Southwest     2534      N232WN  0         E               
227469 Southwest      902      N724SW  0         E               
227470 Southwest     1167      N455WN  0         E               
227471 Southwest      513      N362SW  0         E               
227472 Southwest      581      N646SW  0         E               
227473 Southwest     1223      N394SW  0         E               
227474 Southwest     3085      N510SW  0         E               
227475 Southwest       39      N754SW  0         E               
227476 Southwest      424      N769SW  0         E               
227477 Southwest     1098      N448WN  0         E               
227478 Southwest     1403      N430WN  0         E               
227479 Southwest       33      N698SW  0         E               
227480 Southwest     1212      N226WN  0         E               
227481 Southwest      207      N354SW  0         E               
227482 Southwest      405      N617SW  0         E               
227483 Southwest      628      N389SW  0         E               
227484 Southwest      994      N713SW  0         E               
227485 Southwest     1231      N284WN  0         E               
227486 Southwest      124      N522SW  0         E               
227487 Southwest      280      N728SW  0         E               
227488 Southwest      782      N476WN  0         E               
227489 Southwest     1050      N655WN  0         E               
227490 Southwest      201      N903WN  0         E               
227491 Southwest      471      N632SW  0         E               
227492 Southwest     1191      N284WN  0         E               
227493 Southwest     1674      N366SW  0         E               
227494 Southwest      127      N777QC  0         E               
227495 Southwest      621      N727SW  0       

In [18]:
# Print out a tbl as described in the third instruction, using only helper functions.
select(hflights, contains(c('Time')), contains(c('Delay')))

DepTime ArrTime ActualElapsedTime AirTime ArrDelay DepDelay
1      1400    1500    60                40      -10       0      
2      1401    1501    60                45       -9       1      
3      1352    1502    70                48       -8      -8      
4      1403    1513    70                39        3       3      
5      1405    1507    62                44       -3       5      
6      1359    1503    64                45       -7      -1      
7      1359    1509    70                43       -1      -1      
8      1355    1454    59                40      -16      -5      
9      1443    1554    71                41       44      43      
10     1443    1553    70                45       43      43      
11     1429    1539    70                42       29      29      
12     1419    1515    56                41        5      19      
13     1358    1501    63                44       -9      -2      
14     1357    1504    67                47       -6      -3      
15     1359    1459    60                44      -11      -1      
16     1359    1509    70                41       -1      -1      
17     1530    1634    64                48       84      90      
18     1408    1508    60                42       -2       8      
19     1356    1503    67                46       -7      -4      
20     1507    1622    75                42       72      67      
21     1357    1459    62                47      -11      -3      
22     1355    1456    61                44      -14      -5      
23     1356    1501    65                40       -9      -4      
24     1356    1513    77                43        3      -4      
25     1352    1452    60                40      -18      -8      
26     1353    1455    62                40      -15      -7      
27     1356    1458    62                40      -12      -4      
28     1359    1505    66                46       -5      -1      
29     1355    1455    60                46      -15      -5      
30     1359    1456    57                39      -14      -1      
...    ...     ...     ...               ...     ...      ...     
227467  752     857     65                50       2      -3      
227468  651     746     55                43      -9      -4      
227469 1728    1932    244               231       7      23      
227470 1115    1315    240               225     -15       5      
227471 1753    1911     78                59      31      33      
227472 1222    1335     73                61     -10      -3      
227473  836     946     70                58      -9       1      
227474 1352    1749    177               163      59      72      
227475 1850    2046    176               157      71      70      
227476  707     903    176               162      -2      -3      
227477 1335    1528    173               159      18      20      
227478 1005    1158    173               163      13      15      
227479 1723    1845    202               192      70      78      
227480 1139    1304    205               193      -6      -1      
227481 2023    2109     46                38      29      43      
227482 1542    1637     55                43      12      12      
227483 1728    1825     57                39       5       3      
227484 1252    1349     57                43       9       7      
227485  857     955     58                44       0      -3      
227486 1939    2119    100                81      14      39      
227487  556     745    109                87     -10      -4      
227488 1026    1208    102                86     -12       1      
227489 1611    1746     95                83      -9      16      
227490  758    1051    113               100      -4      -2      
227491 1307    1600    113                98       0       7      
227492 1818    2111    113                97      -9       8      
227493 2047    2334    107                94       4       7      
227494  912    1031     79                61      -4  

### Comparison to base R

In [19]:
ex1r <- hflights[c("TaxiIn","TaxiOut","Distance")]
ex1d <- select(hflights, contains(c('Taxi')), Distance)

ex2r <- hflights[c("Year","Month","DayOfWeek","DepTime","ArrTime")]
ex2d <- select(hflights, Year:ArrTime, -3)

ex3r <- hflights[c("TailNum","TaxiIn","TaxiOut")]
ex3d <- select(hflights, TailNum, contains(c('Taxi')))

## Section 4 - The second of five verbs: mutate

### Mutating is creating

In [20]:
# Add the new variable ActualGroundTime to a copy of hflights and save the result as g1.
g1 <- mutate(hflights, ActualGroundTime = ActualElapsedTime - AirTime)

# Add the new variable GroundTime to a g1. Save the result as g2.
g2 <- mutate(g1, GroundTime = TaxiIn + TaxiOut)

# Add the new variable AverageSpeed to g2. Save the result as g3.
g3 <- mutate(g2, AverageSpeed = Distance / AirTime * 60)

# Print out g3
glimpse(g3)

Observations: 227,496
Variables: 25
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15...
$ DayOfWeek         <int> 6, 7, 1, 2, 3, 4, 5, 6, 7, 1, 2, 3, 4, 5, 6, 7, 1...
$ DepTime           <int> 1400, 1401, 1352, 1403, 1405, 1359, 1359, 1355, 1...
$ ArrTime           <int> 1500, 1501, 1502, 1513, 1507, 1503, 1509, 1454, 1...
$ UniqueCarrier     <chr> "American", "American", "American", "American", "...
$ FlightNum         <int> 428, 428, 428, 428, 428, 428, 428, 428, 428, 428,...
$ TailNum           <chr> "N576AA", "N557AA", "N541AA", "N403AA", "N492AA",...
$ ActualElapsedTime <int> 60, 60, 70, 70, 62, 64, 70, 59, 71, 70, 70, 56, 6...
$ AirTime           <int> 40, 45, 48, 39, 44, 45, 43, 40, 41, 45, 42, 41, 4...
$ ArrDelay          <int> -10, -9, -8, 3, -3, -7, -1, -16, 44, 43, 29, 5, -...
$ DepDelay      

## Add multiple variables using mutate

In [21]:
# hflights and dplyr are ready, are you?

# Add a second variable loss_percent to the dataset: m1
m1 <- mutate(hflights, 
        loss = ArrDelay - DepDelay,
        loss_percent = (ArrDelay - DepDelay)/DepDelay * 100
        )
glimpse(m1)

Observations: 227,496
Variables: 24
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15...
$ DayOfWeek         <int> 6, 7, 1, 2, 3, 4, 5, 6, 7, 1, 2, 3, 4, 5, 6, 7, 1...
$ DepTime           <int> 1400, 1401, 1352, 1403, 1405, 1359, 1359, 1355, 1...
$ ArrTime           <int> 1500, 1501, 1502, 1513, 1507, 1503, 1509, 1454, 1...
$ UniqueCarrier     <chr> "American", "American", "American", "American", "...
$ FlightNum         <int> 428, 428, 428, 428, 428, 428, 428, 428, 428, 428,...
$ TailNum           <chr> "N576AA", "N557AA", "N541AA", "N403AA", "N492AA",...
$ ActualElapsedTime <int> 60, 60, 70, 70, 62, 64, 70, 59, 71, 70, 70, 56, 6...
$ AirTime           <int> 40, 45, 48, 39, 44, 45, 43, 40, 41, 45, 42, 41, 4...
$ ArrDelay          <int> -10, -9, -8, 3, -3, -7, -1, -16, 44, 43, 29, 5, -...
$ DepDelay      

In [22]:
# Copy and adapt the previous command to reduce redendancy: m2
m2 <- mutate(hflights, 
        loss = ArrDelay - DepDelay,
        loss_percent = loss/DepDelay * 100
        )
glimpse(m2)

Observations: 227,496
Variables: 24
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15...
$ DayOfWeek         <int> 6, 7, 1, 2, 3, 4, 5, 6, 7, 1, 2, 3, 4, 5, 6, 7, 1...
$ DepTime           <int> 1400, 1401, 1352, 1403, 1405, 1359, 1359, 1355, 1...
$ ArrTime           <int> 1500, 1501, 1502, 1513, 1507, 1503, 1509, 1454, 1...
$ UniqueCarrier     <chr> "American", "American", "American", "American", "...
$ FlightNum         <int> 428, 428, 428, 428, 428, 428, 428, 428, 428, 428,...
$ TailNum           <chr> "N576AA", "N557AA", "N541AA", "N403AA", "N492AA",...
$ ActualElapsedTime <int> 60, 60, 70, 70, 62, 64, 70, 59, 71, 70, 70, 56, 6...
$ AirTime           <int> 40, 45, 48, 39, 44, 45, 43, 40, 41, 45, 42, 41, 4...
$ ArrDelay          <int> -10, -9, -8, 3, -3, -7, -1, -16, 44, 43, 29, 5, -...
$ DepDelay      

In [23]:
# Add the three variables as described in the third instruction: m3
m3 <- mutate(hflights, 
        TotalTaxi = TaxiIn + TaxiOut,
        ActualGroundTime = ActualElapsedTime - AirTime,
        Diff = TotalTaxi - ActualGroundTime
        )
glimpse(m3)

Observations: 227,496
Variables: 25
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15...
$ DayOfWeek         <int> 6, 7, 1, 2, 3, 4, 5, 6, 7, 1, 2, 3, 4, 5, 6, 7, 1...
$ DepTime           <int> 1400, 1401, 1352, 1403, 1405, 1359, 1359, 1355, 1...
$ ArrTime           <int> 1500, 1501, 1502, 1513, 1507, 1503, 1509, 1454, 1...
$ UniqueCarrier     <chr> "American", "American", "American", "American", "...
$ FlightNum         <int> 428, 428, 428, 428, 428, 428, 428, 428, 428, 428,...
$ TailNum           <chr> "N576AA", "N557AA", "N541AA", "N403AA", "N492AA",...
$ ActualElapsedTime <int> 60, 60, 70, 70, 62, 64, 70, 59, 71, 70, 70, 56, 6...
$ AirTime           <int> 40, 45, 48, 39, 44, 45, 43, 40, 41, 45, 42, 41, 4...
$ ArrDelay          <int> -10, -9, -8, 3, -3, -7, -1, -16, 44, 43, 29, 5, -...
$ DepDelay      

## Filter and arrange

### Section 5 - The third of five verbs: filter Logical operators

In [24]:
# All flights that traveled 3000 miles or more
filter(hflights, Distance >= 3000) %>% glimpse()

Observations: 527
Variables: 22
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 1...
$ DayOfWeek         <int> 1, 7, 6, 5, 4, 3, 2, 1, 7, 6, 5, 4, 3, 2, 1, 7, 6...
$ DepTime           <int> 924, 925, 1045, 1516, 950, 944, 924, 1144, 926, 9...
$ ArrTime           <int> 1413, 1410, 1445, 1916, 1344, 1350, 1337, 1605, 1...
$ UniqueCarrier     <chr> "Continental", "Continental", "Continental", "Con...
$ FlightNum         <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ TailNum           <chr> "N69063", "N76064", "N69063", "N77066", "N76055",...
$ ActualElapsedTime <int> 529, 525, 480, 480, 474, 486, 493, 501, 489, 478,...
$ AirTime           <int> 492, 493, 459, 463, 455, 471, 473, 464, 466, 465,...
$ ArrDelay          <int> 23, 20, 55, 326, -6, 0, -13, 135, -15, -10, -16, ...
$ DepDelay          

In [25]:
# All flights flown by one of JetBlue, Southwest, or Delta
filter(hflights, UniqueCarrier %in% c('JetBlue', 'Southwest', 'Delta')) %>% glimpse()

Observations: 48,679
Variables: 22
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 7, 7, 8, 9, 9, 1...
$ DayOfWeek         <int> 6, 6, 7, 7, 1, 1, 2, 2, 3, 3, 4, 5, 5, 6, 7, 7, 1...
$ DepTime           <int> 654, 1639, 703, 1604, 659, 1801, 654, 1608, 700, ...
$ ArrTime           <int> 1124, 2110, 1113, 2040, 1100, 2200, 1103, 2034, 1...
$ UniqueCarrier     <chr> "JetBlue", "JetBlue", "JetBlue", "JetBlue", "JetB...
$ FlightNum         <int> 620, 622, 620, 622, 620, 622, 620, 622, 620, 624,...
$ TailNum           <chr> "N324JB", "N324JB", "N324JB", "N324JB", "N229JB",...
$ ActualElapsedTime <int> 210, 211, 190, 216, 181, 179, 189, 206, 183, 190,...
$ AirTime           <int> 181, 188, 172, 176, 166, 165, 168, 175, 167, 166,...
$ ArrDelay          <int> 5, 61, -6, 31, -19, 111, -16, 25, -14, -6, -17, 0...
$ DepDelay       

In [26]:
# All flights where taxiing took longer than flying
filter(hflights, (TaxiIn + TaxiOut) > AirTime) %>% glimpse()

Observations: 1,389
Variables: 22
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 24, 30, 24, 10, 31, 31, 31, 31, 30, 30, 30, 30, 3...
$ DayOfWeek         <int> 1, 7, 1, 1, 1, 1, 1, 1, 7, 7, 7, 7, 7, 7, 7, 3, 3...
$ DepTime           <int> 731, 1959, 1621, 941, 1301, 2113, 1434, 900, 1304...
$ ArrTime           <int> 904, 2132, 1749, 1113, 1356, 2215, 1539, 1006, 14...
$ UniqueCarrier     <chr> "American", "American", "American", "American", "...
$ FlightNum         <int> 460, 533, 1121, 1436, 241, 1533, 1541, 1583, 241,...
$ TailNum           <chr> "N545AA", "N455AA", "N484AA", "N591AA", "N14629",...
$ ActualElapsedTime <int> 93, 93, 88, 92, 55, 62, 65, 66, 64, 84, 80, 70, 7...
$ AirTime           <int> 42, 43, 43, 45, 27, 30, 30, 32, 31, 40, 37, 30, 3...
$ ArrDelay          <int> 29, 12, 4, 48, -2, 20, 15, 10, 10, 54, 16, 15, 30...
$ DepDelay        

### Combining tests using boolean operators

In [27]:
# All flights that departed before 5am or arrived after 10pm
filter(hflights, DepTime < 500 | ArrTime > 2200) %>% glimpse()

Observations: 27,799
Variables: 22
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 4, 14, 10, 26, 30, 9, 31, 31, 31, 31, 31, 31, 31,...
$ DayOfWeek         <int> 2, 5, 1, 3, 7, 7, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DepTime           <int> 2100, 2119, 1934, 1905, 1856, 1938, 1919, 2116, 1...
$ ArrTime           <int> 2207, 2229, 2235, 2211, 2209, 2228, 2231, 2344, 2...
$ UniqueCarrier     <chr> "American", "American", "American", "American", "...
$ FlightNum         <int> 533, 533, 1294, 1294, 1294, 731, 190, 209, 250, 2...
$ TailNum           <chr> "N4XGAA", "N549AA", "N3BXAA", "N3BXAA", "N3CPAA",...
$ ActualElapsedTime <int> 67, 70, 121, 126, 133, 290, 132, 268, 141, 134, 1...
$ AirTime           <int> 42, 45, 107, 111, 108, 253, 107, 256, 121, 119, 1...
$ ArrDelay          <int> 47, 69, 80, 56, 54, 78, -12, -15, -18, -10, -12, ...
$ DepDelay       

In [28]:
# All flights that departed late but arrived ahead of schedule
filter(hflights, DepDelay > 0 & ArrDelay < 0) %>% glimpse()

Observations: 27,712
Variables: 22
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 2, 5, 18, 18, 12, 13, 26, 1, 10, 12, 15, 17, 27, ...
$ DayOfWeek         <int> 7, 3, 2, 2, 3, 4, 3, 6, 1, 3, 6, 1, 4, 7, 6, 5, 1...
$ DepTime           <int> 1401, 1405, 1408, 721, 2015, 2020, 2009, 1631, 16...
$ ArrTime           <int> 1501, 1507, 1508, 827, 2113, 2116, 2103, 1736, 17...
$ UniqueCarrier     <chr> "American", "American", "American", "American", "...
$ FlightNum         <int> 428, 428, 428, 460, 533, 533, 533, 1121, 1121, 11...
$ TailNum           <chr> "N557AA", "N492AA", "N507AA", "N558AA", "N555AA",...
$ ActualElapsedTime <int> 60, 62, 60, 66, 58, 56, 54, 65, 61, 68, 64, 72, 6...
$ AirTime           <int> 45, 44, 42, 46, 39, 44, 39, 37, 41, 44, 48, 51, 4...
$ ArrDelay          <int> -9, -3, -2, -8, -7, -4, -17, -9, -5, -6, -9, -1, ...
$ DepDelay       

In [29]:
# All cancelled weekend flights
filter(hflights, Cancelled == 1 & DayOfWeek %in% c(6,7)) %>% glimpse()

Observations: 585
Variables: 22
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 9, 29, 9, 9, 9, 2, 29, 9, 1, 9, 9, 9, 9, 8, 9, 9,...
$ DayOfWeek         <int> 7, 6, 7, 7, 7, 7, 6, 7, 6, 7, 7, 7, 7, 6, 7, 7, 7...
$ DepTime           <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ ArrTime           <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ UniqueCarrier     <chr> "American", "Continental", "Continental", "Delta"...
$ FlightNum         <int> 1820, 408, 755, 8, 6726, 1629, 1590, 5229, 298, 2...
$ TailNum           <chr> "N4XCAA", "", "", "N933DL", "N779SK", "N749SW", "...
$ ActualElapsedTime <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ AirTime           <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ ArrDelay          <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ DepDelay          

In [30]:
# All flights that were cancelled after being delayed
filter(hflights, DepDelay > 0 & Cancelled == 1) %>% glimpse()

Observations: 40
Variables: 22
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 2, 2, 2, 2, 2, 3, 4, 4, 4, 4, 4, 4, 4...
$ DayofMonth        <int> 26, 11, 19, 7, 4, 8, 2, 9, 1, 31, 4, 8, 21, 4, 4,...
$ DayOfWeek         <int> 3, 2, 3, 5, 5, 2, 3, 3, 2, 4, 1, 5, 4, 1, 1, 1, 1...
$ DepTime           <int> 1926, 1100, 1811, 2028, 1638, 1057, 802, 904, 150...
$ ArrTime           <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ UniqueCarrier     <chr> "Continental", "US_Airways", "ExpressJet", "Expre...
$ FlightNum         <int> 310, 944, 2376, 3050, 1121, 408, 2189, 2605, 5812...
$ TailNum           <chr> "N77865", "N452UW", "N15932", "N15912", "N537AA",...
$ ActualElapsedTime <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ AirTime           <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ ArrDelay          <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ DepDelay          <

## Blend together what you’ve learned!

In [31]:
# Select the flights that had JFK as their destination: c1
c1 <- filter(hflights, Dest == 'JFK')

# Combine the Year, Month and DayofMonth variables to create a Date column: c2
c2 <- mutate(c1, Date = paste(Year,Month,DayofMonth, sep="-"))

# Print out a selection of columns of c2
select(c2, Date, DepTime, ArrTime, TailNum)

Date       DepTime ArrTime TailNum
1   2011-1-1    654    1124    N324JB 
2   2011-1-1   1639    2110    N324JB 
3   2011-1-2    703    1113    N324JB 
4   2011-1-2   1604    2040    N324JB 
5   2011-1-3    659    1100    N229JB 
6   2011-1-3   1801    2200    N206JB 
7   2011-1-4    654    1103    N267JB 
8   2011-1-4   1608    2034    N267JB 
9   2011-1-5    700    1103    N708JB 
10  2011-1-5   1544    1954    N644JB 
11  2011-1-6   1532    1943    N641JB 
12  2011-1-7    654    1117    N641JB 
13  2011-1-7   1542    1956    N564JB 
14  2011-1-8    654    1058    N630JB 
15  2011-1-9    653    1059    N599JB 
16  2011-1-9   1618    2057    N625JB 
17  2011-1-10   656    1102    N625JB 
18  2011-1-10  1554    2001    N504JB 
19  2011-1-11   653    1053    N504JB 
20  2011-1-11    NA      NA    N537JB 
21  2011-1-12  1532    1953    N504JB 
22  2011-1-13  1522    1938    N597JB 
23  2011-1-14   808    1229    N597JB 
24  2011-1-14  1534    2015    N729JB 
25  2011-1-15   700    1114    N503JB 
26  2011-1-16   652    1055    N706JB 
27  2011-1-16  1551    2004    N565JB 
28  2011-1-17   730    1135    N523JB 
29  2011-1-17  1531    1946    N779JB 
30  2011-1-18   659    1102    N779JB 
... ...        ...     ...     ...    
666 2011-12-17  641    1038    N247JB 
667 2011-12-17 1616    2033    N247JB 
668 2011-12-18  637    1050    N324JB 
669 2011-12-18 1726    2149    N247JB 
670 2011-12-19  640    1106    N247JB 
671 2011-12-19 1543    2004    N292JB 
672 2011-12-20  647    1059    N292JB 
673 2011-12-20 1624    2044    N306JB 
674 2011-12-21  634    1039    N267JB 
675 2011-12-21 1756    2230    N316JB 
676 2011-12-22  634    1039    N231JB 
677 2011-12-22 1640    2053    N306JB 
678 2011-12-23  633    1038    N284JB 
679 2011-12-23 1606    2015    N273JB 
680 2011-12-24  635    1035    N323JB 
681 2011-12-24 1600    2012    N328JB 
682 2011-12-25  638    1052    N273JB 
683 2011-12-25 1608    2013    N192JB 
684 2011-12-26  633    1046    N216JB 
685 2011-12-26 1550    2002    N216JB 
686 2011-12-27  634    1040    N339JB 
687 2011-12-27 1955      39    N294JB 
688 2011-12-28  632    1047    N231JB 
689 2011-12-28 1748    2147    N281JB 
690 2011-12-29  633    1050    N274JB 
691 2011-12-29 1541    1959    N206JB 
692 2011-12-30  634    1040    N317JB 
693 2011-12-30 1541    1957    N296JB 
694 2011-12-31  831    1251    N247JB 
695 2011-12-31 1544    2009    N323JB

## Recap on select, mutate and filter

In [32]:
hflights %>%
  filter(
    Distance > 1000,
    DayOfWeek > 5,
    TaxiIn + TaxiOut < 15
    ) %>%
  glimpse()

Observations: 1,739
Variables: 22
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 23, 30, 30, 29, 23, 23, 23, 22, 16, 16, 16, 16, 1...
$ DayOfWeek         <int> 7, 7, 7, 6, 7, 7, 7, 6, 7, 7, 7, 7, 7, 7, 7, 7, 6...
$ DepTime           <int> 1535, 851, 2234, 1220, 847, 1224, 931, 942, 848, ...
$ ArrTime           <int> 1933, 1230, 2, 1353, 1213, 1345, 1045, 1340, 1136...
$ UniqueCarrier     <chr> "JetBlue", "Continental", "Continental", "Contine...
$ FlightNum         <int> 624, 1058, 1717, 1620, 1058, 1629, 1723, 1, 309, ...
$ TailNum           <chr> "N599JB", "N39726", "N38417", "N87512", "N16709",...
$ ActualElapsedTime <int> 178, 159, 208, 153, 146, 201, 194, 478, 288, 156,...
$ AirTime           <int> 164, 145, 195, 139, 134, 188, 181, 465, 275, 143,...
$ ArrDelay          <int> -27, -13, 89, 19, -30, -27, -28, -10, 12, -14, -1...
$ DepDelay        

In this dataset it is 1,739 flights
In the class the answer was 155
I think they just have the data filtered to one city (Houston)

## Section 6 - Almost there: the arrange verb

### Arranging your data

In [33]:
# Definition of dtc
dtc <- filter(hflights, Cancelled == 1, !is.na(DepDelay))

# Arrange dtc by departure delays
arrange(dtc, DepDelay) %>% glimpse()

Observations: 68
Variables: 22
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 7, 1, 12, 10, 7, 9, 2, 5, 1, 1, 2, 3, 4, 4, 5, 6,...
$ DayofMonth        <int> 23, 17, 1, 12, 29, 29, 9, 9, 20, 17, 21, 18, 30, ...
$ DayOfWeek         <int> 6, 1, 4, 3, 5, 4, 3, 1, 4, 1, 1, 5, 6, 7, 1, 1, 7...
$ DepTime           <int> 605, 916, 541, 2022, 1424, 1639, 555, 715, 1413, ...
$ ArrTime           <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ UniqueCarrier     <chr> "Frontier", "ExpressJet", "US_Airways", "American...
$ FlightNum         <int> 225, 3068, 282, 3724, 1079, 2062, 3265, 1177, 552...
$ TailNum           <chr> "N912FR", "N13936", "N840AW", "N539MQ", "N14628",...
$ ActualElapsedTime <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ AirTime           <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ ArrDelay          <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ DepDelay          <

In [34]:
# Arrange dtc so that cancellation reasons are grouped
arrange(dtc, CancellationCode) %>% glimpse()

Observations: 68
Variables: 22
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 2, 2, 2, 2, 2, 3, 4, 4, 4, 4, 4, 4, 4, 4, 5...
$ DayofMonth        <int> 20, 7, 4, 8, 1, 21, 9, 18, 4, 8, 21, 4, 11, 7, 30...
$ DayOfWeek         <int> 4, 5, 5, 2, 2, 1, 3, 5, 1, 5, 4, 1, 1, 4, 6, 7, 1...
$ DepTime           <int> 1413, 2028, 1638, 1057, 1508, 2257, 555, 727, 163...
$ ArrTime           <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ UniqueCarrier     <chr> "United", "ExpressJet", "American", "Continental"...
$ FlightNum         <int> 552, 3050, 1121, 408, 5812, 1111, 3265, 109, 8, 4...
$ TailNum           <chr> "N509UA", "N15912", "N537AA", "N11641", "N959SW",...
$ ActualElapsedTime <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ AirTime           <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ ArrDelay          <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ DepDelay          <

In [35]:
# Arrange dtc according to carrier and departure delays
arrange(dtc, UniqueCarrier, DepDelay) %>% glimpse()

Observations: 68
Variables: 22
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 6, 8, 2, 10, 2, 7, 4, 4, 5, 9, 5, 9, 7, 1, 8, 7, ...
$ DayofMonth        <int> 11, 18, 4, 12, 9, 17, 30, 10, 23, 29, 16, 26, 29,...
$ DayOfWeek         <int> 6, 4, 5, 3, 3, 7, 6, 7, 1, 4, 1, 1, 5, 3, 4, 1, 3...
$ DepTime           <int> 1649, 1808, 1638, 2022, 555, 1917, 612, 1147, 657...
$ ArrTime           <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ UniqueCarrier     <chr> "AirTran", "American", "American", "American_Eagl...
$ FlightNum         <int> 1595, 1294, 1121, 3724, 3265, 3717, 5386, 5402, 5...
$ TailNum           <chr> "N946AT", "N3FLAA", "N537AA", "N539MQ", "N613MQ",...
$ ActualElapsedTime <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ AirTime           <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ ArrDelay          <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ DepDelay          <

In [36]:
# Arrange dtc according to carrier and departure delays
arrange(dtc, UniqueCarrier, DepDelay) %>% glimpse()

Observations: 68
Variables: 22
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 6, 8, 2, 10, 2, 7, 4, 4, 5, 9, 5, 9, 7, 1, 8, 7, ...
$ DayofMonth        <int> 11, 18, 4, 12, 9, 17, 30, 10, 23, 29, 16, 26, 29,...
$ DayOfWeek         <int> 6, 4, 5, 3, 3, 7, 6, 7, 1, 4, 1, 1, 5, 3, 4, 1, 3...
$ DepTime           <int> 1649, 1808, 1638, 2022, 555, 1917, 612, 1147, 657...
$ ArrTime           <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ UniqueCarrier     <chr> "AirTran", "American", "American", "American_Eagl...
$ FlightNum         <int> 1595, 1294, 1121, 3724, 3265, 3717, 5386, 5402, 5...
$ TailNum           <chr> "N946AT", "N3FLAA", "N537AA", "N539MQ", "N613MQ",...
$ ActualElapsedTime <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ AirTime           <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ ArrDelay          <int> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N...
$ DepDelay          <

In [37]:
# Arrange flights by total delay (normal order).
arrange(hflights, (DepDelay + ArrDelay)) %>% glimpse()

Observations: 227,496
Variables: 22
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 7, 8, 8, 8, 8, 12, 1, 8, 8, 8, 8, 9, 12, 9, 12, 8...
$ DayofMonth        <int> 3, 31, 21, 28, 29, 25, 30, 3, 4, 18, 26, 11, 24, ...
$ DayOfWeek         <int> 7, 3, 7, 7, 1, 7, 7, 3, 4, 4, 5, 7, 6, 2, 6, 2, 7...
$ DepTime           <int> 1914, 934, 935, 2059, 935, 741, 620, 1741, 930, 9...
$ ArrTime           <int> 2039, 1039, 1039, 2206, 1041, 926, 812, 1810, 104...
$ UniqueCarrier     <chr> "ExpressJet", "SkyWest", "SkyWest", "SkyWest", "S...
$ FlightNum         <int> 2804, 2040, 2001, 2003, 2040, 4591, 4461, 2603, 1...
$ TailNum           <chr> "N12157", "N783SK", "N767SK", "N783SK", "N767SK",...
$ ActualElapsedTime <int> 85, 185, 184, 187, 186, 165, 172, 89, 191, 184, 1...
$ AirTime           <int> 66, 172, 171, 171, 169, 147, 156, 73, 177, 172, 1...
$ ArrDelay          <int> -70, -56, -56, -54, -54, -57, -49, -40, -49, -52,...
$ DepDelay      

In [38]:
# Keep flights leaving to DFW before 8am and arrange according to decreasing AirTime 
hflights %>% filter(Dest == 'DFW', DepTime < 800) %>% arrange(desc(AirTime)) %>% glimpse()

Observations: 799
Variables: 22
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 11, 8, 10, 5, 4, 4, 6, 9, 3, 12, 4, 4, 5, 6, 6, 1...
$ DayofMonth        <int> 22, 25, 12, 2, 4, 4, 21, 1, 14, 5, 25, 11, 11, 21...
$ DayOfWeek         <int> 2, 4, 3, 1, 1, 1, 2, 4, 1, 1, 1, 1, 3, 2, 1, 7, 2...
$ DepTime           <int> 635, 602, 559, 716, 741, 627, 726, 715, 729, 724,...
$ ArrTime           <int> 825, 758, 738, 854, 949, 742, 848, 844, 917, 847,...
$ UniqueCarrier     <chr> "American", "American_Eagle", "American_Eagle", "...
$ FlightNum         <int> 1903, 3265, 3265, 2237, 1225, 3265, 2259, 1948, 1...
$ TailNum           <chr> "N477AA", "N633MQ", "N632MQ", "N552AA", "N4XVAA",...
$ ActualElapsedTime <int> 110, 116, 99, 98, 128, 75, 82, 89, 108, 83, 91, 7...
$ AirTime           <int> 81, 74, 71, 70, 63, 62, 62, 62, 61, 61, 59, 59, 5...
$ ArrDelay          <int> 40, 53, 33, 29, 89, 37, 9, 9, 33, 2, 20, 1, 0, 18...
$ DepDelay          

## Summarise and the pipe operator

## Section 7 - Last but not least: summarise The syntax of summarize

In [39]:
glimpse(hflights)

Observations: 227,496
Variables: 22
$ Year              <int> 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2011, 2...
$ Month             <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1...
$ DayofMonth        <int> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15...
$ DayOfWeek         <int> 6, 7, 1, 2, 3, 4, 5, 6, 7, 1, 2, 3, 4, 5, 6, 7, 1...
$ DepTime           <int> 1400, 1401, 1352, 1403, 1405, 1359, 1359, 1355, 1...
$ ArrTime           <int> 1500, 1501, 1502, 1513, 1507, 1503, 1509, 1454, 1...
$ UniqueCarrier     <chr> "American", "American", "American", "American", "...
$ FlightNum         <int> 428, 428, 428, 428, 428, 428, 428, 428, 428, 428,...
$ TailNum           <chr> "N576AA", "N557AA", "N541AA", "N403AA", "N492AA",...
$ ActualElapsedTime <int> 60, 60, 70, 70, 62, 64, 70, 59, 71, 70, 70, 56, 6...
$ AirTime           <int> 40, 45, 48, 39, 44, 45, 43, 40, 41, 45, 42, 41, 4...
$ ArrDelay          <int> -10, -9, -8, 3, -3, -7, -1, -16, 44, 43, 29, 5, -...
$ DepDelay      

In [40]:
# Print out a summary with variables min_dist and max_dist
summarize(hflights, min_dist = min(Distance), max_dist = max(Distance))

min_dist max_dist
1 79       3904

In [41]:
# Print out a summary with variable max_div
hflights %>% filter(Diverted == 1) %>% summarize(max_div = max(Distance))

max_div
1 3904

## Aggregate functions

Aggregate functions defined in R:

min(x) - minimum value of vector x
max(x) - maximum value of vector x
mean(x) - mean value of vector x
median(x) - median value of vector x
quantile(x, p) - pth quantile of vector x
sd(x) - standard deviation of vector x
var(x) - variance of vector x
IQR(x) - Inter Quartile Range (IQR) of vector x
diff(range(x)) - total range of vector x

In [42]:
# Remove rows that have NA ArrDelay: temp1
temp1 <- filter(hflights, !is.na(ArrDelay))

# Generate summary about ArrDelay column of temp1
summarize(temp1, 
  earliest = min(ArrDelay),
  average = mean(ArrDelay),
  latest = max(ArrDelay),
  sd = sd(ArrDelay)
  )

earliest average  latest sd      
1 -70      7.094334 978    30.70852

In [43]:
# Keep rows that have no NA TaxiIn and no NA TaxiOut: temp2
temp2 <- filter(hflights, !is.na(TaxiIn) & !is.na(TaxiOut))

# Print the maximum taxiing difference of temp2 with summarise()
summarise(temp2, max_taxi_diff = max(abs(TaxiIn - TaxiOut)))

max_taxi_diff
1 160

## dplyr aggregate functions

dplyr has some of its own aggregate functions:

first(x) - The first element of vector x
last(x) - The last element of vector x
nth(x, n) - The nth element of vector x
n() - The number of rows in the data.frame or group of observations that summarise() describes
n_distinct(x) - The number of unique values in vector x

In [44]:
# Generate summarizing statistics for hflights
summarise(hflights, 
  n_obs = n(),
  n_carrier = n_distinct(UniqueCarrier),
  n_dest = n_distinct(Dest),
  dest100 = nth(Dest, 100)
  )

n_obs  n_carrier n_dest dest100
1 227496 15        116    DFW

In [46]:
# Filter hflights to keep all American Airline flights: aa
aa <- filter(hflights, UniqueCarrier == 'American')

# Generate summarizing statistics for aa 
summarise(aa, 
  n_flights = n(),
  n_canc = sum(Cancelled),
  p_canc = n_canc/n_flights * 100,
  avg_delay = mean(ArrDelay, na.rm=T)
  )

n_flights n_canc p_canc   avg_delay
1 3244      60     1.849568 0.8917558

## Section 8 - Chaining your functions: the pipe operator

the pipe operator %>% is probably my favorite thing in R
it takes the output from the last function and passes it into the next function as the first argument
But you can also use a . to pass it into the next function in any location!
This lets you make code easily readable from left to right and top to bottom.
I can’t say enough about how much better this makes it to read code and see what someone is doing

In [47]:
# Write the 'piped' version of the English sentences.
hflights %>%
  mutate(diff = TaxiOut - TaxiIn) %>%
  filter(!is.na(diff)) %>%
  summarize(avg = mean(diff))

avg     
1 8.992064

### Drive of fly? Part 1 of 2

In [48]:
# Build data frame with 4 columns of hflights and 2 self-defined columns: d
d <- hflights %>%
  select(Dest, UniqueCarrier, Distance, ActualElapsedTime) %>%
  mutate(
    RealTime = ActualElapsedTime + 100,
    mph = Distance/RealTime*60
    )

# Filter and summarise d according to the instructions
d %>%
  filter(
    !is.na(mph),
    mph < 70
    ) %>%
  summarize(
    n_less = n(),
    n_dest = n_distinct(Dest),
    min_dist = min(Distance),
    max_dist = max(Distance)
    )

n_less n_dest min_dist max_dist
1 6726   13     79       305

### Drive or fly? Part 2 of 2

In [50]:
# Solve the exercise using a combination of dplyr verbs and %>%
hflights %>%
  mutate(RealTime = ActualElapsedTime + 100, mph = Distance / RealTime * 60) %>%
  filter(mph < 105 | Cancelled == 1 | Diverted == 1) %>%
  summarise(n_non = n(), 
            p_non = n_non / nrow(hflights) * 100, 
            n_dest = n_distinct(Dest), 
            min_dist = min (Distance), 
            max_dist = max(Distance))

n_non p_non    n_dest min_dist max_dist
1 42400 18.63769 113    79       3904

## Advanced piping exercise

In [51]:
hflights %>%
  filter(
    !is.na(ArrTime),
    !is.na(DepTime),
    ArrTime < DepTime
    ) %>%
  summarise(n = n())

n   
1 2718

## Group_by and working with databases

## Section 9 - get group-wise insights: group_by

Combining group_by with summarize is very powerful
You can also combine it with mutate and arrange to create powerful window functions

### Unite and conquer using group_by

In [52]:
# Make an ordered per-carrier summary of hflights
hflights %>%
 group_by(UniqueCarrier) %>%
 summarise(
   n_flights = n(),
   n_canc = sum(Cancelled),
   p_canc = n_canc/n_flights * 100,
   avg_delay = mean(ArrDelay, na.rm=T)
   ) %>%
 arrange(avg_delay, p_canc)

UniqueCarrier      n_flights n_canc p_canc    avg_delay 
1  US_Airways          4082       46   1.1268986 -0.6307692
2  American            3244       60   1.8495684  0.8917558
3  AirTran             2139       21   0.9817672  1.8536239
4  Alaska               365        0   0.0000000  3.1923077
5  Mesa                  79        1   1.2658228  4.0128205
6  Delta               2641       42   1.5903067  6.0841374
7  Continental        70032      475   0.6782614  6.0986983
8  American_Eagle      4648      135   2.9044750  7.1529751
9  Atlantic_Southeast  2204       76   3.4482759  7.2569543
10 Southwest          45343      703   1.5504047  7.5871430
11 Frontier             838        6   0.7159905  7.6682692
12 ExpressJet         73053     1132   1.5495599  8.1865242
13 SkyWest            16061      224   1.3946828  8.6934922
14 JetBlue              695       18   2.5899281  9.8588410
15 United              2072       34   1.6409266 10.4628628

In [53]:
# Make an ordered per-day summary of hflights
hflights %>%
  group_by(DayOfWeek) %>%
  summarize(avg_taxi = mean(TaxiIn + TaxiOut, na.rm=T)) %>%
  arrange(desc(avg_taxi))

DayOfWeek avg_taxi
1 1         21.77027
2 2         21.43505
3 4         21.26076
4 3         21.19055
5 5         21.15805
6 7         20.93726
7 6         20.43061

### Combine group_by with mutate

In [54]:
# Solution to first instruction
hflights %>%
  filter(!is.na(ArrDelay)) %>%
  group_by(UniqueCarrier) %>%
  summarize(p_delay = sum(ArrDelay > 0)/n()) %>%
  mutate(rank = rank(p_delay)) %>%
  arrange(rank)

UniqueCarrier      p_delay   rank
1  American           0.3030208  1  
2  AirTran            0.3112269  2  
3  US_Airways         0.3267990  3  
4  Atlantic_Southeast 0.3677511  4  
5  American_Eagle     0.3696714  5  
6  Delta              0.3871092  6  
7  JetBlue            0.3952452  7  
8  Alaska             0.4368132  8  
9  Southwest          0.4644557  9  
10 Mesa               0.4743590 10  
11 Continental        0.4907385 11  
12 ExpressJet         0.4943420 12  
13 United             0.4963109 13  
14 SkyWest            0.5350105 14  
15 Frontier           0.5564904 15

In [55]:
# Solution to second instruction
hflights %>%
  filter(
    !is.na(ArrDelay),
    ArrDelay > 0
    ) %>%
  group_by(UniqueCarrier) %>%
  summarize(avg = mean(ArrDelay)) %>%
  mutate(rank = rank(avg)) %>%
  arrange(rank)

UniqueCarrier      avg      rank
1  Mesa               18.67568  1  
2  Frontier           18.68683  2  
3  US_Airways         20.70235  3  
4  Continental        22.13374  4  
5  Alaska             22.91195  5  
6  SkyWest            24.14663  6  
7  ExpressJet         24.19337  7  
8  Southwest          25.27750  8  
9  AirTran            27.85693  9  
10 American           28.49740 10  
11 Delta              32.12463 11  
12 United             32.48067 12  
13 American_Eagle     38.75135 13  
14 Atlantic_Southeast 40.24231 14  
15 JetBlue            45.47744 15

### Advanced group_by exercises

In [56]:
# Which plane (by tail number) flew out of Houston the most times? How many times? adv1
adv1 <- hflights %>%
  group_by(TailNum) %>%
  summarize(n = n()) %>%
  filter(n == max(n))
adv1

TailNum n  
1 N14945  971

In [57]:
# How many airplanes only flew to one destination from Houston? adv2
adv2 <- hflights %>%
  group_by(TailNum) %>%
  summarize(n_dest = n_distinct(Dest)) %>%
  filter(n_dest == 1) %>%
  summarize(nplanes = n())
adv2

nplanes
1 1526

In [58]:
# Find the most visited destination for each carrier: adv3
adv3 <- hflights %>%
  group_by(UniqueCarrier, Dest) %>%
  summarize(n = n()) %>%
  group_by(UniqueCarrier) %>%
  mutate(rank = rank(desc(n))) %>%
  filter(rank == 1) %>%
  arrange(UniqueCarrier, rank)
adv3

UniqueCarrier      Dest n    rank
1  AirTran            ATL  2029 1   
2  Alaska             SEA   365 1   
3  American           DFW  2105 1   
4  American_Eagle     DFW  2424 1   
5  Atlantic_Southeast DTW   851 1   
6  Continental        EWR  3924 1   
7  Delta              ATL  2396 1   
8  ExpressJet         CRP  3175 1   
9  Frontier           DEN   837 1   
10 JetBlue            JFK   695 1   
11 Mesa               CLT    71 1   
12 SkyWest            COS  1335 1   
13 Southwest          DAL  8243 1   
14 United             SFO   643 1   
15 US_Airways         CLT  2212 1

In [59]:
# Find the carrier that travels to each destination the most: adv4
adv4 <- hflights %>%
  group_by(Dest, UniqueCarrier) %>%
  summarize(n = n()) %>%
  group_by(Dest) %>%
  mutate(rank = rank(desc(n))) %>%
  filter(rank == 1)
adv4  

Dest UniqueCarrier n    rank
1   ABQ  Southwest     1019 1   
2   AEX  ExpressJet     724 1   
3   AGS  Continental      1 1   
4   AMA  ExpressJet    1297 1   
5   ANC  Continental    125 1   
6   ASE  SkyWest        125 1   
7   ATL  Delta         2396 1   
8   AUS  Continental   2645 1   
9   AVL  ExpressJet     350 1   
10  BFL  SkyWest        504 1   
11  BHM  ExpressJet    2040 1   
12  BKG  AirTran        110 1   
13  BNA  ExpressJet    2104 1   
14  BOS  Continental   1752 1   
15  BPT  ExpressJet       3 1   
16  BRO  ExpressJet    1692 1   
17  BTR  ExpressJet    1762 1   
18  BWI  Continental   1324 1   
19  CAE  ExpressJet     561 1   
20  CHS  ExpressJet     906 1   
21  CID  ExpressJet     410 1   
22  CLE  Continental   2140 1   
23  CLT  US_Airways    2212 1   
24  CMH  ExpressJet     964 1   
25  COS  SkyWest       1335 1   
26  CRP  ExpressJet    3175 1   
27  CRW  ExpressJet     357 1   
28  CVG  ExpressJet    1178 1   
29  DAL  Southwest     8243 1   
30  DAY  ExpressJet     451 1   
... ...  ...           ...  ... 
87  PBI  ExpressJet     809 1   
88  PDX  Continental   1235 1   
89  PHL  Continental   1610 1   
90  PHX  Continental   1725 1   
91  PIT  Continental    837 1   
92  PNS  ExpressJet    1539 1   
93  PSP  SkyWest        106 1   
94  RDU  ExpressJet     821 1   
95  RIC  ExpressJet     900 1   
96  RNO  Continental    243 1   
97  RSW  Continental    636 1   
98  SAN  Continental   2225 1   
99  SAT  Continental   2692 1   
100 SAV  ExpressJet     863 1   
101 SDF  ExpressJet    1269 1   
102 SEA  Continental   2250 1   
103 SFO  Continental   2175 1   
104 SHV  ExpressJet     787 1   
105 SJC  Continental    885 1   
106 SJU  Continental    391 1   
107 SLC  SkyWest       1251 1   
108 SMF  Continental   1014 1   
109 SNA  Continental   1661 1   
110 STL  Southwest     1334 1   
111 TPA  Continental   1725 1   
112 TUL  ExpressJet    1574 1   
113 TUS  SkyWest       1032 1   
114 TYS  ExpressJet    1210 1   
115 VPS  ExpressJet     880 1   
116 XNA  ExpressJet    1172 1

## dplyr deals with different types

In [61]:
library(data.table)

# Convert hflights to a data.table
class(hflights)

[1] "tbl_df"     "tbl"        "data.frame"

In [62]:
hflights2 <- as.data.table(hflights)
class(hflights2)

[1] "data.table" "data.frame"

In [63]:
# Use summarise to calculate n_carrier
s2 <- hflights2 %>%
  summarize(n_carrier = n_distinct(UniqueCarrier))
s2

n_carrier
1 15